sdk/src/azure/core/az_json_writer.c

```diff
@@ -782,10 +782,12 @@ az_json_writer_append_json_text(az_json_writer* ref_json_writer, az_span json_te
   az_span remaining_json = _get_remaining_span(ref_json_writer, _az_MINIMUM_STRING_CHUNK_SIZE);
   _az_RETURN_IF_NOT_ENOUGH_SIZE(remaining_json, _az_MINIMUM_STRING_CHUNK_SIZE);
 
+  int32_t required_size = az_span_size(json_text);
   if (ref_json_writer->_internal.need_comma)
   {
     remaining_json = az_span_copy_u8(remaining_json, ',');
     ref_json_writer->_internal.bytes_written++;
+    required_size++; // For the leading comma separator.
```

Comment:
Can you add some unit tests for this, that showcase the error before the fix?

Reply:
If you only use publicly exposed APIs in unit tests, the issue isn't visible to the end user:

```C
 {
      TEST_EXPECT_SUCCESS(az_json_writer_init(&writer, AZ_SPAN_FROM_BUFFER(array), NULL));
      TEST_EXPECT_SUCCESS(az_json_writer_append_begin_array(&writer));
      TEST_EXPECT_SUCCESS(az_json_writer_append_json_text(&writer, AZ_SPAN_FROM_STR("1")));
      TEST_EXPECT_SUCCESS(az_json_writer_append_json_text(&writer, AZ_SPAN_FROM_STR("2")));
      TEST_EXPECT_SUCCESS(az_json_writer_append_json_text(&writer, AZ_SPAN_FROM_STR("3")));
      TEST_EXPECT_SUCCESS(az_json_writer_append_end_array(&writer));

      az_span_to_str((char*)array, 200, az_json_writer_get_bytes_used_in_destination(&writer));
      assert_int_equal(7, az_span_size(az_json_writer_get_bytes_used_in_destination(&writer)));
      assert_string_equal(array, "[1,2,3]");
    }
```

Reply:
That may be the case that the unit test does not pick this up. But we picked it up when using the API from the Azure RTOS IoT middleware's `nx_azure_iot_json_writer_append_json_text()` function. When using this SDK via Azure RTOS IoT middleware, the resulting NX_PACKETs do hold wrong length information. So without the PR some internal data structure is not updated correctly. You also can see by reviewing the code that the additional comma is not taken into account in the same way as it it with the other append functions. All other append functions take into account the additional comma when calling `_az_update_json_writer_state()`.

Reply:
I agree there is a correctness issue here :)
Ideally, the upstream dependencies do not use internal state like this directly, since we don't guarantee to not change or break the internal state. In fact, this particular field was only there (atm) for testing only:
https://github.com/Azure/azure-sdk-for-c/blob/9b5ef3d889b014d6e1b777d80f495dae27fe3ae1/sdk/inc/azure/core/az_json.h#L311

Reply:
> Can you add some regression unit tests for this, that showcase the error before the fix?

Added a test here using the internals, just to avoid regression in the future: https://github.com/Azure/azure-sdk-for-c/pull/1907